counter/src/main/java/org/infinispan/counter/impl/CounterModuleLifecycle.java

```diff
@@ -0,0 +1,86 @@
+import org.infinispan.counter.impl.listener.CounterFilterAndConvert;
+import org.infinispan.counter.impl.listener.CounterSnapshot;
+import org.infinispan.counter.impl.weak.WeakCounterKey;
+import org.infinispan.factories.GlobalComponentRegistry;
+import org.infinispan.lifecycle.AbstractModuleLifecycle;
+import org.infinispan.lifecycle.ModuleLifecycle;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.transaction.TransactionMode;
+import org.kohsuke.MetaInfServices;
+
+import java.util.Map;
+
+/**
+ * It register a {@link CounterManager} to each {@link EmbeddedCacheManager} started and starts the cache on it.
+ *
+ * @author Pedro Ruivo
+ * @since 9.0
+ */
+@MetaInfServices(value = ModuleLifecycle.class)
+public class CounterModuleLifecycle extends AbstractModuleLifecycle {
+
+   private static Configuration createConfiguration() {
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      defineClustering(builder);
+      disableTransactions(builder);
+      return builder.build();
+   }
+
+   private static void defineClustering(ConfigurationBuilder builder) {
+      builder.clustering().cacheMode(CacheMode.DIST_SYNC);
+      builder.clustering().hash().numOwners(2);
+      builder.clustering().l1().disable(); //bad for write intensive!
+   }
+
+   private static void disableTransactions(ConfigurationBuilder builder) {
+      builder.transaction().transactionMode(TransactionMode.NON_TRANSACTIONAL);
+   }
+
+   private static void addAdvancedExternalizer(Map<Integer, AdvancedExternalizer<?>> map, AdvancedExternalizer<?> ext) {
+      map.put(ext.getId(), ext);
+   }
+
+   @Override
+   public void cacheManagerStarting(GlobalComponentRegistry gcr, GlobalConfiguration globalConfiguration) {
+      final Map<Integer, AdvancedExternalizer<?>> externalizerMap = globalConfiguration.serialization().advancedExternalizers();
+
+      addAdvancedExternalizer(externalizerMap, new AddFunction.AddFunctionExternalizer());
+      addAdvancedExternalizer(externalizerMap, new ResetFunction.ResetFunctionExternalizer());
+      addAdvancedExternalizer(externalizerMap, new SimpleCounterValue.Externalizer());
+      addAdvancedExternalizer(externalizerMap, new ThresholdCounterValue.ThresholdCounterEntryExternalizer());
+      addAdvancedExternalizer(externalizerMap, CounterFilterAndConvert.EXTERNALIZER);
+      addAdvancedExternalizer(externalizerMap, CounterSnapshot.EXTERNALIZER);
+      addAdvancedExternalizer(externalizerMap, AtomicCounterKey.EXTERNALIZER);
+      addAdvancedExternalizer(externalizerMap, WeakCounterKey.EXTERNALIZER);
+   }
+
+   @Override
+   public void cacheManagerStarted(GlobalComponentRegistry gcr) {
+      final EmbeddedCacheManager cacheManager = gcr.getComponent(EmbeddedCacheManager.class);
+      if (!cacheManager.getCacheConfigurationNames().contains(CounterManager.COUNTER_CACHE_NAME)) {
+         cacheManager.defineConfiguration(CounterManager.COUNTER_CACHE_NAME, createConfiguration());
```

Comment:
It would probably make sense to introduce volatile and persistent counters. For this we should make the cache persistent (easily done via InternalCacheRegistry flags) and then use the SKIP_CACHE_STORE/LOAD accordingly